Better validation logic for workflows

mephenor opened the pull request:
This PR introduces a NoOpValidator class that is used as  a dummy to skip validation in a workflow setting.
The new default for transformations is to skip validation if not explicitly instructed to validate.
Validation of input and output can be configured separately.

There's no strong reason for making validation opt-in, apart from the solution being slightly more compact.
If there are strong arguments to keep the validation on by default, the logic can be changed easily.

This also bumps the version to 5.0.0, so we can draft a release based on schemapack.

mephenor committed:
Added skip validation logic

mephenor committed:
Template file update

mephenor committed:
Update major version for release

sbilge commented on src/metldata/transform/handling.py:
```diff
@@ -27,6 +34,26 @@
+    ):
+        """Do nothing with the input, just skip plugin creation."""
+
+    @override
+    def validate(self, *, datapack: DataPack):
+        """Do nothing and return."""
+        return
```
My environment gives `Method "validate" is marked as an override, but no base method was found with this nameMypy[misc](https://mypy.readthedocs.io/en/latest/_refs.html#code-misc)` error.

mephenor replied:
Can't reproduce.
Can you check if your environment has the correct mypy version, i.e. rebuild the devcontainer, if necessary?

sbilge reviewed (changes requested):
I only noticed a mypy complaint.

Apart from that, I know we currently don’t have negative test cases, but shall we add one here to ensure that a validation error is raised when a datapack is not compatible with its schemapack, both at the beginning and at the end of the process?

Alternatively, we could add an integration test that produces an incompatible intermediate datapack–schemapack pair, although I’m not entirely sure how to simulate that. Perhaps we could use a mock transformation that intentionally creates non-compatible data–model pairs.

mephenor committed:
Workflow tests for invalid input/output models

mephenor committed:
Template updates

sbilge commented on tests/workflow/test_validation.py:
```diff
@@ -0,0 +1,42 @@
+
+@pytest.mark.parametrize("test_case", VALIDATION_TEST_CASES, ids=str)
+def test_workflow_invalid_models(test_case: WorkflowTestCase):
+    """Test the happy path of running a workflow."""
```
Sounds quite unhappy to me :)

mephenor replied:
Fixed the desciption

sbilge commented on tests/workflow/test_validation.py:
```diff
@@ -0,0 +1,42 @@
+        else PostTransformValidationError
+    )
+    with pytest.raises(exception_type):
+        _ = handler.run(data=test_case.input_data, annotation=test_case.annotation)
```
```suggestion
        handler.run(data=test_case.input_data, annotation=test_case.annotation)
```

mephenor replied:
Wanted to be explicit about this not using the return value, but can simplify it

sbilge replied:
I think it is clearer that you are not using a return value when you don't assign it to something, because it is already inside of `pytest.raises`

sbilge commented on tests/fixtures/workflow.py:
```diff
@@ -144,9 +147,21 @@ def _read_test_case(
+    """Read validation test cases."""
+    return [
+        _read_test_case(workflow_dir=WORKFLOW_VALIDATION_DIR, case_name=case_name)
+        for case_name in VALIDATION_WORKFLOWS
```
We should create the validation test case differently since those test cases do not include a comparison between the expected and observed transformation results.

mephenor replied:
Done, the transformed_model/transformed_data fields can now be None, skipping loading based on the value of `load_transformed`

sbilge commented on tests/fixtures/workflow_validation/invalid_input_model/workflow.yaml:
```diff
@@ -0,0 +1,135 @@
+operations:
```
Let's switch to a much simpler workflow for this test case

sbilge commented on tests/fixtures/workflow_validation/invalid_output_model/workflow.yaml:
```diff
@@ -0,0 +1,134 @@
+operations:
```
Let's switch to a simpler workflow

sbilge commented on tests/fixtures/workflow_validation/invalid_input_model/transformed.datapack.yaml:
We should modify how we create the test cases and delete the unnecessary transformed.* files that do not contribute to the test case.

sbilge commented on tests/fixtures/workflow_validation/invalid_input_model/transformed.schemapack.yaml:
See above

sbilge commented on tests/fixtures/workflow_validation/invalid_output_model/transformed.datapack.yaml:
See above

sbilge commented on tests/fixtures/workflow_validation/invalid_output_model/transformed.schemapack.yaml:
See above

sbilge commented on tests/fixtures/workflow_validation/invalid_input_model/workflow.yaml:
Let's switch to a much simpler workflow

sbilge commented on tests/fixtures/workflow_validation/invalid_output_model/workflow.yaml:
Let's switch to a much simpler workflow

mephenor committed:
Adjusted transformed test output

mephenor committed:
Bug fix + additional test cases

mephenor committed:
Correct test cases

mephenor committed:
Ignore all the types

mephenor commented:
> Alternatively, we could add an integration test that produces an incompatible intermediate datapack–schemapack pair, although I’m not entirely sure how to simulate that. Perhaps we could use a mock transformation that intentionally creates non-compatible data–model pairs.

Added a fixture to intercept constructor call args to check expectations that way.
Added a test case where the second schemapack is invalid to make sure validation only happens at the end.
Added a test case with a workflow containing only one step to test that edge case.

sbilge commented on tests/workflow/test_validation.py:
```diff
@@ -0,0 +1,90 @@
+
+@pytest.fixture()
+def capture_constructor_args():
+    """Capture call args for TransformationHandler constructor."""
```
It would be nice to have a bit more description here.

mephenor replied:
What's missing here?

mephenor replied:
Added a bit more text.

sbilge commented on tests/workflow/test_validation.py:
```diff
@@ -0,0 +1,90 @@
+}
+
+
+class CallArgs:
```
How about we call this `ArgsContainer` to prevent `call_args.call_args` repetition in the following fixture?

mephenor replied:
Refactored and didn't check... I'll adjust it in a second

sbilge commented on tests/workflow/test_validation.py:
```diff
@@ -0,0 +1,90 @@
+    call_args = CallArgs()
+
+    def capture(self, *args, **kwargs):
+        call_args.call_args.append(
```
See above

sbilge commented on tests/workflow/test_validation.py:
```diff
@@ -0,0 +1,90 @@
+        )
+        self._real_init(*args, **kwargs)
+
+    TransformationHandler._real_init = TransformationHandler.__init__  # type: ignore
+    TransformationHandler.__init__ = capture  # type: ignore
+
+    yield call_args
+
+    TransformationHandler.__init__ = TransformationHandler._real_init  # type: ignore
+    del TransformationHandler._real_init  # type: ignore
```
I think pytest's monkeypatch handles saving/restoring automatically. Did you not use it for a reason?

sbilge replied:
Something like this:

```
@pytest.fixture()
def capture_constructor_args(monkeypatch):
    """Capture call args for TransformationHandler constructor."""
    call_args = CallArgs()

    original = TransformationHandler.__init__

    def capture(self, *args, **kwargs):
        call_args.call_args.append(
            {
                "validate_input": kwargs["validate_input"],
                "validate_output": kwargs["validate_output"],
            }
        )
        return original(self, *args, **kwargs)

    monkeypatch.setattr(TransformationHandler, "__init__", capture)

    return call_args
```

mephenor replied:
This will likely overwrite the original (again), based on when exactly the constructor is actually called.
Had that issue with other solution attempts before.
Also, don't really like to use monkeypatch inside a fixture, makes it less transparent.

mephenor committed:
Requested changes

mephenor committed:
Extended fixture description
